File: algorithms-data-structures/go/src/data-structures/trees/path-sum/sum.go
package pathsum

import "go-ds/src/data-structures/trees"

func hasPathSum(root *trees.TreeNode, targetSum int) bool {
	if root == nil {
		return false
	}
	if root.Val == targetSum && root.Left == nil && root.Right == nil {
		return true
	}
	if hasPathSum(root.Left, targetSum-root.Val) {
		return true
	}
	if hasPathSum(root.Right, targetSum-root.Val) {
		return true
	}
	return false
}

File: algorithms-data-structures/ts/src/data-structures/tree/path-sum/sum.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

import {TreeNode} from "../TreeNode";

export function hasPathSum(root: TreeNode | null, targetSum: number): boolean {
  if (root === null) return false;
  // If we are at a leaf node and the target sum is equal to the value of the leaf node, return true
  if (root.val === targetSum && root.left === null && root.right === null)
    return true;
  // Recursively call the function on the left and right nodes, subtracting the value of the current node from the target sum
  if (hasPathSum(root.left, targetSum - root.val)) return true;
  // Recursively call the function on the right node, subtracting the value of the current node from the target sum
  if (hasPathSum(root.right, targetSum - root.val)) return true;
  // If we reach this point, there is no path that sums to the target sum
  return false;
}
